modules/gateway/upnp.go

```diff
@@ -38,10 +38,15 @@ func myExternalIP() (string, error) {
 	return string(buf[:n-1]), nil
 }
 
-// learnHostname discovers the external IP of the Gateway. Once the IP has
-// been discovered, it registers the ShareNodes RPC to be called on new
+// threadedLearnHostname discovers the external IP of the Gateway. Once the IP
+// has been discovered, it registers the ShareNodes RPC to be called on new
 // connections, advertising the IP to other nodes.
-func (g *Gateway) learnHostname(port string) {
+func (g *Gateway) threadedLearnHostname() {
```

Comment:
@VoidingWarranties and I went back and forth on it a bit. In general, it can be seen as the gateway that's forwarding it's port, which is why it makes sense for this function to be a method of the gateway. This point is hammered home more when you add the logging calls, which necessarily need the gateway's logger.

Since you're already using the gateway, it makes sense to also use the port within the gateway instead of passed as an argument. And since you don't want the gateway to close if this function hasn't shut down.